### PR TITLE
[WIP] Yul to WebAssembly code generator

### PR DESCRIPTION
**Pull request by axic**
Part of #2405.

### CODE REVIEW

File: libsolidity/interface/AssemblyStack.h [hunk 1]
```diff
@@ -63,6 +63,9 @@ class AssemblyStack
 	/// Run the assembly step (should only be called after parseAndAnalyze).
 	eth::LinkerObject assemble(Machine _machine);
 
+	/// Run the assembly step (should only be called after parseAndAnalyze) and print source.
+	std::string assemblySource(Machine _machine);
```
Comment (axic) [inline libsolidity/interface/AssemblyStack.h]: Basically there came up the need to return an assembly source as opposed to bytecode. Perhaps we could introduce `AssemblyMachineObject`, which includes the assembly source and the what `LinkerObject` contains (all the fields are optional).

File: libsolidity/interface/AssemblyStack.cpp [hunk 2]
```diff
@@ -72,12 +73,31 @@ eth::LinkerObject AssemblyStack::assemble(Machine _machine)
...
 		solUnimplemented("EVM 1.5 backend is not yet implemented.");
 	case Machine::eWasm:
-		solUnimplemented("eWasm backend is not yet implemented.");
+		solUnimplemented("eWASM backend is not yet implemented.");
 	}
 	// unreachable
 	return eth::LinkerObject();
 }
 
+string AssemblyStack::assemblySource(Machine _machine)
```
Comment (axic) [inline libsolidity/interface/AssemblyStack.cpp]: We go with `assembly()`. Returns empty string if not supported by the machine.

File: libsolidity/interface/AssemblyStack.cpp [hunk 3]
```diff
@@ -72,12 +73,31 @@ eth::LinkerObject AssemblyStack::assemble(Machine _machine)
 	case Machine::EVM15:
 		solUnimplemented("EVM 1.5 backend is not yet implemented.");
 	case Machine::eWasm:
-		solUnimplemented("eWasm backend is not yet implemented.");
+		solUnimplemented("eWASM backend is not yet implemented.");
```
Comment (axic) [inline libsolidity/interface/AssemblyStack.cpp]: Return empty `LinkerObject()` if not supported by the machine.

File: libsolidity/interface/AssemblyStack.cpp (hunk 2 above)
Comment (axic) [inline libsolidity/interface/AssemblyStack.cpp]: Rename `AssemblyStack` to `BackendStack`.

**Comment by axic** [comment]
This depends on #2361.

**Comment by axic** [comment]
This can generate very primitive (almost compliant) wasm code now:
```
======= wasm.asm (eWasm) =======

Pretty printed source:
{
    function mul(a:u64, b:u64) -> c:u64
    {
    }
    function main() -> y:u64
    {
        let x:u64 := mul(1:u64, 2:u64)
        y := add64(1:u64, 2:u64)
    }
}

Binary representation:
No binary representation found.

Text representation:
(module 
    (func $mul 
        (param $a i64)
        (param $b i64)
        (result i64)
        (local $c i64)
        (return $c)
    )
    (func $main 
        (result i64)
        (local $y i64)
        (local $x i64)
        (set_local $x 
            (call $mul
                 (i64.const 1)
                 (i64.const 2)
            )
        )
        (set_local $y 
            (i64.add 
                (i64.const 1)
                (i64.const 2)
            )
        )
        (return $y)
    )
)
```

Above, `add64` is a builtin function which translates to `i64.add`.

**Comment by axic** [comment]
@chriseth can you review this? The `IndentedWriter` could be merged separately too (it could also move to `libdevcore`).

File: libjulia/backends/webassembly/IndentedWriter.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @date 2017
+ * Julia to WebAssembly code generator.
```
Comment (chriseth) [inline libjulia/backends/webassembly/IndentedWriter.cpp]: JULIA?

File: libjulia/backends/webassembly/IndentedWriter.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	string result;
+	for (auto const& line: m_lines)
+		result += string(line.indentation * 4, ' ') + line.contents + "\n";
+	return result;
+}
+
+void IndentedWriter::newLine()
+{
+	if (!m_lines.back().contents.empty())
+		m_lines.push_back({ "", m_lines.back().indentation });
```
Comment (chriseth) [inline libjulia/backends/webassembly/IndentedWriter.cpp]: This looks like you only need to store the current indentation and the full text up to (excluding) the current line.

File: libjulia/backends/webassembly/IndentedWriter.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @date 2017
+ * Indented text writer.
```
Comment (chriseth) [inline libjulia/backends/webassembly/IndentedWriter.h]: This file is not specific to webassembly. What about `libdevcore`?

File: libjulia/backends/webassembly/WebAssembly.cpp [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+
+#include <memory>
+#include <functional>
+
+using namespace std;
+using namespace dev;
+using namespace dev::solidity;
+using namespace dev::solidity::assembly;
+
+class Generator: public boost::static_visitor<>
```
Comment (chriseth) [inline libjulia/backends/webassembly/WebAssembly.cpp]: Can you move this into a separate file and split header and implementation?

File: libjulia/backends/webassembly/WebAssembly.cpp [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+	{
+		solAssert(false, "Assignment from stack is not supported in Julia.");
+	}
+	void operator()(assembly::Label const&)
+	{
+		solAssert(false, "Labels are not supported in Julia.");
+	}
+	void operator()(assembly::Literal const& _literal)
+	{
+		solAssert(!m_onRoot, "Statements are only allowed within functions.");
```
Comment (chriseth) [inline libjulia/backends/webassembly/WebAssembly.cpp]: This is not true, please change to "unimplemented" assertion.

Comment (chriseth) [reply libjulia/backends/webassembly/WebAssembly.cpp]: Also below.

File: libjulia/backends/webassembly/WebAssembly.cpp [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+
+		output.add("(get_local $" + _identifier.name + ")");
+	}
+	void operator()(assembly::VariableDeclaration const& _varDecl)
+	{
+		solAssert(!m_onRoot, "Statements are only allowed within functions.");
+
+		solAssert(_varDecl.variables.size() == 1, "Tuples not supported yet.");
```
Comment (chriseth) [inline libjulia/backends/webassembly/WebAssembly.cpp]: Please change to "unimplemented" assertion.

File: libjulia/backends/webassembly/WebAssembly.cpp [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+		output.addLine("(set_local $" + _assignment.variableName.name + " ");
+		output.indent();
+		boost::apply_visitor(*this, *_assignment.value);
+		output.unindent();
+		output.add(")");
+		output.newLine();
+	}
+	void operator()(assembly::FunctionDefinition const& _funDef)
+	{
+		solAssert(m_onRoot, "Embededd functions are not supported.");
```
Comment (chriseth) [inline libjulia/backends/webassembly/WebAssembly.cpp]: typo

File: libjulia/backends/webassembly/WebAssembly.cpp [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+		}
+		output.unindent();
+		output.addLine(")");
+	}
+	void operator()(assembly::Switch const& _switch)
+	{
+		solAssert(!m_onRoot, "Statements are only allowed within functions.");
+
+		solAssert(_switch.cases.size() <= 2, "");
+		/// One of the cases must be the default case
```
Comment (chriseth) [inline libjulia/backends/webassembly/WebAssembly.cpp]: There could also be no default case?

File: libjulia/backends/webassembly/WebAssembly.cpp [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+			boost::apply_visitor(*this, _funCall.arguments[0]);
+			output.newLine();
+			boost::apply_visitor(*this, _funCall.arguments[1]);
+			output.unindent();
+			output.add(")");
+			return true;
+		}
+		else if (_funCall.functionName.name == "sub64")
+		{
+			output.add("(i64.sub ");
```
Comment (chriseth) [inline libjulia/backends/webassembly/WebAssembly.cpp]: This looks like code repetition :-)

File: libjulia/backends/webassembly/WebAssembly.cpp [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+			boost::apply_visitor(*this, _funCall.arguments[1]);
+			output.unindent();
+			output.add(")");
+			return true;
+		}
+
+		return false;
+	}
+
+	IndentedWriter output;
```
Comment (chriseth) [inline libjulia/backends/webassembly/WebAssembly.cpp]: Style: `m_output`

File: libjulia/backends/webassembly/WebAssembly.cpp (hunk 11 above)
Comment (axic) [reply libjulia/backends/webassembly/WebAssembly.cpp]: For this version of the implementation it needs it.

File: libjulia/backends/webassembly/WebAssembly.cpp (hunk 8 above)
Comment (axic) [reply libjulia/backends/webassembly/WebAssembly.cpp]: So I imagined to have a transformation outside which moves root level statements into a function. Considering that as a prerequisite to this code, the above is an assert. Considering it is not a prerequisite (the wasm code could be called without the transformation), this is an unimplemented assert.

**Comment by axic** [comment]
Rebased.

**Comment by chriseth** [comment]
As an experiment we could rewrite this from scratch, just ignore types and assume everything is 64 bit. @axic wanted to work on that.

**Comment by stackenbotten** [comment]
There was an error when running `test_check_style` for commit `033b6c5751491ad7541efc32c4f31c00ae6b42cc`:
```
Error: Trailing whitespace found:
 libyul/backends/webassembly/WebAssembly.cpp:106: 

```
Please check that your changes are working as intended.

**Comment by chriseth** [comment]
Superseded by https://github.com/ethereum/solidity/pull/6556

**Pull request closed without merging**